### PR TITLE
Oj 958 get addresses

### PR DESCRIPTION
**Pull request by MattBidewell**
### What changed

Added tests, Infra and code for a new lambda to access the personIdentity table and return any saved addresses.

### Why did it change

As part of the partial addresses, a new lambda to return any presaved addresses associated with the session is needed. This lambda is written in NodeJS and will access the personidentity table and return any addresses associated with the session ID.

### Issue tracking

- [OJ-958](https://govukverify.atlassian.net/browse/OJ-958)

### Environment variables or secrets

- [X] No environment variables or secrets were added or changed

### Other considerations

**This will be done in a separate PR**
- [ ] Update [README](./blob/main/README.md) with any new instructions or tasks

### CODE REVIEW

File: lambdas/get-addresses/jest.config.ts [hunk 1]
```diff
@@ -5,11 +5,11 @@
 
 export default {
     transform: {
-        "^.+\\.ts?$": "esbuild-jest",
+        "^.+\\.ts?$": "ts-jest",
```
Comment (MattBidewell) [inline lambdas/get-addresses/jest.config.ts:8]: use ts-jest to build tests and not esbuild.

esbuild + breakpoints on vscode was acting strange and not breakpointing in the right locations. The log lines were also slightly incorrect for the best output, it was strange.

File: infrastructure/lambda/private-api.yaml [hunk 2]
```diff
@@ -174,6 +174,45 @@ paths:
...
+                $ref: "#/components/schemas/Error"
+        "200":
+          description: "200 response"
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/PostcodeLookupResponse"
+      x-amazon-apigateway-request-validator: "Validate both"
+      x-amazon-apigateway-integration:
+        httpMethod: "POST"
```
Comment (MattBidewell) [inline infrastructure/lambda/private-api.yaml:206]: Is this supposed to be POST

File: lambdas/get-addresses/package.json [hunk 3]
```diff
@@ -1,30 +1,34 @@
...
+    "name": "get_addresses",
     "version": "1.0.0",
-    "description": "hello world sample for NodeJS",
+    "description": "GetAddresses Typescript Node Lambda",
     "main": "app.js",
     "dependencies": {
-        "esbuild": "0.14.14"
+        "esbuild": "0.14.14",
+        "@aws-sdk/client-dynamodb": "3.150.0",
+        "@aws-sdk/lib-dynamodb": "3.150.0"
```
Comment (MattBidewell) [inline lambdas/get-addresses/package.json:9]: esbuild doesnt deploy unless it knows about these types in the dependancies... Its strange as I would assume esbuild is aware that lambda knows that we have the entire SDK

File: infrastructure/lambda/private-api.yaml (hunk 2 above)
Comment (techaman-m) [reply infrastructure/lambda/private-api.yaml:206]: Yes, more details here - https://aws.amazon.com/premiumsupport/knowledge-center/api-gateway-lambda-template-invoke-error/

File: lambdas/get-addresses/src/app.ts [hunk 4]
```diff
@@ -1,34 +1,37 @@
...
 
 export const lambdaHandler = async (event: APIGatewayProxyEvent): Promise<APIGatewayProxyResult> => {
     let response: APIGatewayProxyResult;
     try {
+        const sessionId = event.headers["session_id"] as string;
+
+        if (!sessionId) {
+            response = {
+                statusCode: 400,
+                body: "Missing header: session_id is required",
```
Comment (techaman-m) [inline lambdas/get-addresses/src/app.ts:14]: Do we need this as this validation is already being done here https://github.com/alphagov/di-ipv-cri-address-api/pull/203/files#diff-027c6904962e08ce3c1694bc9a276b0644ba677533f3c15e0a6269d211c8ea03R184

Comment (MattBidewell) [reply lambdas/get-addresses/src/app.ts:14]: I added it for defensive programming purposes.

If we ever got rid of the APIGateway config or it didnt work as intended then there wouldn't be any check on the session ID check

File: lambdas/get-addresses/src/services/address-service.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
+/* eslint-disable no-console */
+import { DynamoDBDocument, GetCommand } from "@aws-sdk/lib-dynamodb";
+import { CanonicalAddress } from "../types/address";
+
+export class AddressService {
+    constructor(private tableName: string, private dynamoDbClient: DynamoDBDocument) {}
+
+    public async getAddressesBySessionId(sessionId: string | undefined): Promise<CanonicalAddress[]> {
+        try {
+            console.log(`TableName: ${this.tableName}\nsessionId: ${sessionId}`);
```
Comment (MattBidewell) [inline lambdas/get-addresses/src/services/address-service.ts:10]: remove

File: lambdas/get-addresses/src/app.ts [hunk 6]
```diff
@@ -1,34 +1,37 @@
...
             body: JSON.stringify({
-                message: "hello world",
+                result,
             }),
         };
-    } catch (err: unknown) {
+    } catch (err) {
         // eslint-disable-next-line no-console
-        console.log(err);
+        console.error(err);
```
Comment (MattBidewell) [inline lambdas/get-addresses/src/app.ts:30]: todo - move to powertools logger

File: lambdas/get-addresses/tests/unit/app.test.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+        const params = {
+            headers: {
+                session_id: "123-abc",
+            },
+        } as unknown as APIGatewayProxyEvent;
+
+        const result = await lambdaHandler(params);
+        const resultAddress = JSON.parse(result.body).result;
+        expect(resultAddress).toEqual(addresses);
```
Comment (techaman-m) [inline lambdas/get-addresses/tests/unit/app.test.ts:34]: Can also do an assert for `sessionId` returned in the payload for this and other tests?
E.g.
`expect(resultAddress.sessionId).toEqual('123-abc');`

File: lambdas/get-addresses/tests/unit/app.test.ts [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+                session_id: "123-abc",
+            },
+        } as unknown as APIGatewayProxyEvent;
+        const result = await lambdaHandler(params);
+        const resultAddress = JSON.parse(result.body).result;
+        expect(resultAddress).toEqual([]);
+        expect(result.statusCode).toBe(200);
+    });
+
+    it("should return a sttaus code 400 when no sesson id is passed", async () => {
```
Comment (techaman-m) [inline lambdas/get-addresses/tests/unit/app.test.ts:51]: typo in word `status`

File: lambdas/get-addresses/tests/unit/app.test.ts (hunk 7 above)
Comment (MattBidewell) [reply lambdas/get-addresses/tests/unit/app.test.ts:34]: I'm not sure that would be needed as we're testing the `session_id` exists in the `GetCommand` in the `address-service.test.ts`. We also dont use the sessionId after its returned so we would be testing something not needed (imo)

**Review by techaman-m — DISMISSED** [review]

File: infrastructure/lambda/private-api.yaml [hunk 9]
```diff
@@ -174,6 +174,45 @@ paths:
           Fn::Sub: arn:${AWS::Partition}:apigateway:${AWS::Region}:lambda:path/2015-03-31/functions/arn:${AWS::Partition}:lambda:${AWS::Region}:${AWS::AccountId}:function:${CommonStackName}-AuthorizationFunction/invocations
         passthroughBehavior: "when_no_match"
 
+  /addresses:
+    get:
+      parameters:
+        - in: header
+          name: "session_id"
```
Comment (MattBidewell) [inline infrastructure/lambda/private-api.yaml:181]: session_id or session-id

File: lambdas/get-addresses/src/app.ts [hunk 10]
```diff
@@ -1,34 +1,36 @@
...
     let response: APIGatewayProxyResult;
     try {
+        const sessionId = event.headers["session_id"] as string;
+        if (!sessionId) {
+            response = {
+                statusCode: 400,
+                body: "Missing header: session_id is required",
+            };
+            return response;
+        }
```
Comment (idrop) [inline lambdas/get-addresses/src/app.ts:16]: could we not do his validation in the APIGW?

**Review by idrop — APPROVED** [review]

File: lambdas/get-addresses/tsconfig.json [hunk 11]
```diff
@@ -3,14 +3,16 @@
         "target": "es2020",
         "strict": true,
         "preserveConstEnums": true,
-        "noEmit": true,
-        "sourceMap": false,
+        // "noEmit": true, // emit for debugger testing purposes.
```
Comment (merlinc) [inline lambdas/get-addresses/tsconfig.json:6]: Does `tsconfig.json` allow comments inside it - normal JSON errors if you try that.

File: .pre-commit-config.yaml [hunk 12]
```diff
@@ -1,14 +1,19 @@
 # See https://pre-commit.com for more information
```
Comment (merlinc) [inline .pre-commit-config.yaml:1]: `pre-commit` <3

File: lambdas/get-addresses/tests/unit/app.test.ts [hunk 13]
```diff
@@ -0,0 +1,72 @@
+import { APIGatewayProxyEvent } from "aws-lambda";
```
Comment (merlinc) [inline lambdas/get-addresses/tests/unit/app.test.ts:1]: The [GDS Way for Node.js](https://gds-way.cloudapps.digital/manuals/programming-languages/nodejs/#project-directory-structure) says:

 > Store test files within the implementation

This is based on the standard JS way of working which doesn't necessarily fit with Typescript.

Thoughts on co-locating of tests vs a separate unit test folder?

File: lambdas/get-addresses/.eslintrc.js [hunk 14]
```diff
@@ -7,6 +7,7 @@ module.exports = {
     env: {
```
Comment (merlinc) [inline lambdas/get-addresses/.eslintrc.js:7]: When we get another typescript lambda, we should move this linting config one folder up.

File: .pre-commit-config.yaml [hunk 15]
```diff
@@ -1,14 +1,19 @@
...
-        types: [file]
\ No newline at end of file
+      - id: eslint
+        files: \.[jt]sx?$ # *.js, *.jsx, *.ts and *.tsx
+        types: [file]
+  - repo: https://github.com/pre-commit/mirrors-prettier
+    rev: "v2.7.1"
+    hooks:
+      - id: prettier
+        types_or: ["javascript", "ts", "json"]
```
Comment (merlinc) [inline .pre-commit-config.yaml:19]: We could also apply prettier to Markdown and Yaml if we wanted.

File: .pre-commit-config.yaml [hunk 16]
```diff
@@ -1,14 +1,19 @@
...
+        args: ["--baseline", ".secrets.baseline"]
+  - repo: https://github.com/pre-commit/mirrors-eslint
+    rev: "v8.24.0" # Use the sha / tag you want to point at
     hooks:
-    -   id: eslint
-        files: \.[jt]sx?$  # *.js, *.jsx, *.ts and *.tsx
-        types: [file]
\ No newline at end of file
+      - id: eslint
+        files: \.[jt]sx?$ # *.js, *.jsx, *.ts and *.tsx
```
Comment (merlinc) [inline .pre-commit-config.yaml:13]: Is this plugin automatically checking directories, or is this configured to only test for files in the root folder?

**Review by merlinc — APPROVED** [review]
Looks good to me, tests seem to cover all the options.

I have a number of non-blocking comments, but anything they bring up can be a future refactoring

**Pull request merged**